packages/toolpad-components/src/Autocomplete.tsx

```diff
@@ -0,0 +1,150 @@
+import { SX_PROP_HELPER_TEXT } from './constants.js';
+
+type AutocompleteOption = string | { label?: string; value?: string };
+type AutocompleteValue = string | null;
+
+interface AutocompleteProps
+  extends Omit<
+    MuiAutocompleteProps<AutocompleteOption, false, false, false>,
+    'renderInput' | 'value' | 'onChange'
+  > {
+  value: AutocompleteValue;
+  onChange: (newValue: AutocompleteValue) => void;
+  options: AutocompleteOption[];
+  labelField?: string;
+  valueField?: string;
+  label: string;
+}
+
+function Autocomplete({
+  options,
+  labelField,
+  valueField,
+  label,
+  onChange,
+  value,
+  ...rest
+}: AutocompleteProps) {
+  const [selectedVal, setSelectedVal] = React.useState<AutocompleteOption | null>(null);
+
+  const getValue = React.useCallback(
+    (selection: AutocompleteOption | null): AutocompleteValue => {
+      if (!selection) {
+        return null;
+      }
+      if (typeof selection === 'string') {
+        return selection;
+      }
+      if (typeof selection === 'object') {
+        if (valueField) {
+          return (selection as any)?.[valueField];
+        }
+        return selection?.value ?? selection?.label ?? null;
+      }
+      return null;
+    },
+    [valueField],
+  );
+
+  const getOptionLabel = React.useCallback(
+    (option: AutocompleteOption) => {
+      if (!option) {
+        return '';
+      }
+      if (typeof option === 'string') {
+        return option;
+      }
+      if (typeof option === 'object') {
+        if (labelField) {
+          return (option as any)?.[labelField];
+        }
+        return option?.label ?? '';
+      }
+      return '';
+    },
+    [labelField],
+  );
+
+  const handleChange = React.useCallback(
+    (event: React.SyntheticEvent<Element>, selection: AutocompleteOption | null) => {
+      const newValue: AutocompleteValue = getValue(selection);
+      onChange(newValue);
+      setSelectedVal(selection);
+    },
+    [onChange, getValue],
+  );
+  return (
+    <MuiAutocomplete
+      onChange={handleChange}
+      options={options ?? []}
+      isOptionEqualToValue={(option, selectedValue) => getValue(option) === getValue(selectedValue)}
+      getOptionLabel={getOptionLabel}
+      value={selectedVal}
+      renderInput={(params) => <TextField {...params} label={label} variant="outlined" />}
+      {...rest}
+    />
+  );
+}
+
+export default createComponent(Autocomplete, {
+  layoutDirection: 'both',
+  argTypes: {
+    options: {
+      helperText: 'The options available to search from.',
+      type: 'array',
+      default: [],
+      control: {
+        type: 'SelectOptions',
+      },
+    },
+    value: {
+      helperText: 'The value of the autocomplete.',
+      type: 'string',
+      onChangeProp: 'onChange',
+      default: '',
+    },
+    label: {
+      helperText: 'The label to display for the autocomplete.',
+      type: 'string',
+      default: 'Search…',
+    },
+    labelField: {
```

Comment:
Why is this needed? Can we just start with expecting from users to supply correct `AutocompleteOption`?

Reply:
Without this, we wouldn't be able to support the Autocomplete docs examples; a lot of the times, the data is in a particular shape and being able to specify which field is to act as the `label` and which is to be the `value` is a very useful feature. I would do it very frequently with React Admin; in fact the example in the video is what I would spend a lot of time doing again and again on multiple admin tools through code.

Reply:
> Without this, we wouldn't be able to support the use case where the data is in a particular shape and being able to specify which field is to act as the label and which is to be the value is needed.

I believe you could bind to:
```tsx
customOptions.map(option => ({ label: option.notLabel, value: option.notvalue }))
```

> Without valueField, I would first have to transform the data I am connecting to Toolpad, and only then use the Autocomplete with it.

Yes, I propose we start with that use case

Reply:
Got it; I'll remove them and add these as a proposal in the Product discussion notes